Add class methods to encode integer as bit vector representation, as …

junyechen1996 opened the pull request:
…a vector of field elements.

- It's a common functionality used by both Prio3Sum and Prio3SumVec, and increases readability IMO.

cjpatton reviewed (commented):
Nice! A couple minor comments.

junyechen1996 commented on draft-irtf-cfrg-vdaf.md:
```diff
@@ -1673,6 +1673,48 @@ def decode_vec(Field, encoded: Bytes) -> Vec[Field]:
+
+~~~
+def encode_into_bit_vector(Field,
+                           val: Unsigned,
```
It is a bit odd. The reason is when we encode integer into bit vector, the measurement is always in `Unsigned` form. When we `truncate`, we need to decode the vector into a `Field`, because the `Field` is actually secret-shared. I could instead make the function name more explicit, e.g. `encode_integer_into_bit_vector` and `decode_field_from_bit_vector`?

cjpatton replied:
Hmm, fair. I think I'd still prefer to take in a `Field`, even if that means we need to  `Field.encode_into_bit_vector(elem=Field(23))` then call `elem.as_unsigned()` in the encoding method internally. But I don't feel strongly about it. Feel free to leave as-is, and we'll see if maybe @divergentdave has a strong opinion here.

junyechen1996 commented on draft-irtf-cfrg-vdaf.md:
```diff
@@ -1673,6 +1673,48 @@ def decode_vec(Field, encoded: Bytes) -> Vec[Field]:
+    for l in range(bits):
+        encoded.append(Field(x & 1))
+        x >>= 1
+    if x != 0:
```
Done.

cjpatton reviewed (approved):
Looks reasonable to me. @junyechen1996 please squash and fix-up the commit message before merging

junyechen1996 commented:
Got it. I squashed and fixed the commit message, and updated PR title and description according to the commit message. Will wait for more reviews in the meantime.

cjpatton reviewed (commented):
One more suggestion here: It would be more consistent with our API if the encoder returned the encoded value rather than appended the encoded value to a buffer. Something like this:
* `Field.encode_into_bit_vector(val: Unsigned, bits: Unsinged) -> list[Field]`
* `Field.decoded_from_bit_vector(encoded: list[Field]) -> Field`

junyechen1996 commented:
> One more suggestion here: It would be more consistent with our API if the encoder returned the encoded value rather than appended the encoded value to a buffer. Something like this:
> 
> * `Field.encode_into_bit_vector(val: Unsigned, bits: Unsinged) -> list[Field]`
> * `Field.decoded_from_bit_vector(encoded: list[Field]) -> Field`

Makes sense. Initially I made `encode_into_bit_vector` take in the vector so we don't have to allocate it for every coordinate in `Prio3SumVec`, but I guess we could just use python list concatenation.

I kept the method name `decode_from_bit_vector`, I'm not sure if `decoded_from_bit_vector` is what you intended?

cjpatton reviewed (approved):
Thanks @junyechen1996 !
